### PR TITLE
Licensing Portal: Add company details form

### PR DESCRIPTION
**Pull request by kallehauge**
#### Summary

This PR aims to add a new page to the Licensing Portal where partners can update their company details like company name or company address, so they have up-to-date information on their invoices.

#### Changes proposed in this Pull Request

* Add a new page Company Details page (`/partner-portal/company-details`).

##### Context

* 1201902532330136-as-1201955149001159
* pdpAdu-bj-p2

##### Submit new data

* Apply D79508-code and point your DNS to your sandbox
* Spin up the branch
* Go to `http://jetpack.cloud.localhost:3000/partner-portal/company-details`
* Add address information (could be [Automattics address](https://automattic.com/contact/))
* Submit the form and (hopefully) you should now see a success message
* Hard refresh the page to check if the data was persistent

##### Tracking

* Verify that your submission shows up in tracks as a `calypso_partner_portal_update_company_details_submit` event (there might be a 10min delay): 2c4d2-pb

#### Screenshot

Screenshot 2022-05-19 at 11 06 51

### CODE REVIEW

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	const dispatch = useDispatch();
+	// The partner has already been fetched through e.g. the requireAccessContext
+	// controller function, so we do not have to do any further checks.
+	const partner = useSelector( getCurrentPartner );
+	const { countryOptions, stateOptions } = useCountriesAndStates();
+	const submitNotificationId = 'company-details-form';
+	const showCountryFields = countryOptions.length > 0;
+
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const [ name, setName ] = useState( castAsString( partner?.name ) );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Not a blocker: `castAsString` seems a bit overkill as `?? ''` should suffice for these values.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+						label: value,
+						isLabel: false,
+					};
+				} else {
+					// The countries/states endpoint uses the following formats if they include states:
+					// * Code: US:TX
+					// * Label: United States (US) — Texas
+					// so we wish to split the code by ":" and labels by " — ".
+					const { 0: countryCode, 1: stateCode } = key.split( ':' );
+					const { 0: countryLabel, 1: stateLabel } = value.split( ' — ' );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts]: Shouldn't this be just `const [ countryCode, stateCode ] = key.split( ... )`?
It probably works because arrays are objects but let's not abuse this unnecessarily :D

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+export function useCountriesAndStates() {
+	const { data: countriesList } = useCountries();
+
+	return useMemo( () => {
+		const countryOptions = {};
+		const stateOptions = {};
+
+		if ( countriesList ) {
+			Object.entries( countriesList ).map( ( [ key, value ] ) => {
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts]: We can reduce the indentation here for readability by dropping the if() and replacing it with `Object.entries( countriesList ?? [] )`

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				// We just have to add the country to the list of countries if the key / country
+				// code doesn't include a colon since that means they don't have any states.
+				if ( key.indexOf( ':' ) === -1 ) {
+					countryOptions[ key ] = {
+						value: key,
+						label: value,
+						isLabel: false,
+					};
+				} else {
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts]: We can reduce the indentation for readability of a big chunk of the function by adding an early return instead of an else clause.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import { useMemo } from 'react';
+import { useCountries } from 'calypso/landing/stepper/hooks/use-countries';
+
+/**
+ * List of countries Stripe allow state/area/province values for.
+ */
+const stripeSupportedStateCountries = [
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts:8]: It's unfortunate that Stripe.js doesn't provide this 😩

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	const [ city, setCity ] = useState( castAsString( partner?.address.city ) );
+	const [ line1, setLine1 ] = useState( castAsString( partner?.address.line1 ) );
+	const [ line2, setLine2 ] = useState( castAsString( partner?.address.line2 ) );
+	const [ postalCode, setPostalCode ] = useState( castAsString( partner?.address.postal_code ) );
+	const [ addressState, setAddressState ] = useState( castAsString( partner?.address.state ) );
+
+	const updateCompanyDetails = useUpdateCompanyDetailsMutation( {
+		onSuccess: () => {
+			dispatch(
+				successNotice( translate( 'Company details has been updated' ), {
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]:
```suggestion
				successNotice( translate( 'Company details have been updated' ), {
```

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+import './style.scss';
+
+export default function CompanyDetailsForm(): ReactElement {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	// The partner has already been fetched through e.g. the requireAccessContext
+	// controller function, so we do not have to do any further checks.
+	const partner = useSelector( getCurrentPartner );
+	const { countryOptions, stateOptions } = useCountriesAndStates();
+	const submitNotificationId = 'company-details-form';
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Let's be more specific as notice IDs are global for Calypso Green - we usually prefix the partner portal ones with `partner-portal-`.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	const showCountryFields = countryOptions.length > 0;
+
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: This is not necessary - we can use `updateCompanyDetails.isLoading` which useMutation provides.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+		updateCompanyDetails.mutate( payload );
+
+		dispatch(
+			recordTracksEvent( 'calypso_partner_portal_update_company_details_submit', {
+				partner_id: partner?.id,
+				...payload,
+			} )
+		);
+	}, [ dispatch, updateCompanyDetails ] );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Let's add `submitNotificationId` to the dependency list. Even if it doesn't change right now it is being used and if it becomes dynamic this callback will break in a hidden way.

Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Let's depend on `updateCompanyDetails.mutate` instead of the entire `updateCompanyDetails` object because we only care about `updateCompanyDetails.mutate` (which likely will never change) while the whole object probably changes every render.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+	return (
+		<Card className="company-details-form">
+			<FormFieldset>
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Let's use a proper form element and:
- handle "onSubmit" on it instead of the button
- disable form submission until showCountryFields is true so users don't submit the form without the country fields.

File: client/jetpack-cloud/sections/partner-portal/primary/company-details-dashboard/index.tsx [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+		<Main wideLayout className="company-details-dashboard">
+			<DocumentHead title={ translate( 'Company Details' ) } />
+			<SidebarNavigation />
+
+			<div className="company-details-dashboard__header">
+				<CardHeading size={ 36 }>{ translate( 'Company Details' ) }</CardHeading>
+
+				<p className="company-details-dashboard__description">
+					{ translate(
+						'Update your company details. Updates will appear on all future receipts.'
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/primary/company-details-dashboard/index.tsx]: How about this wording:
```suggestion
						'Update your company details. Changes will be applied to all future invoices.'
```

File: client/state/partner-portal/partner/hooks/use-update-company-details.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import wpcom from 'calypso/lib/wp';
+import { APIError, APIPartner } from 'calypso/state/partner-portal/types';
+
+interface MutationUpdateCompanyDetailsVariables {
+	name?: string;
+	country?: string;
+	city?: string;
+	line1?: string;
+	line2?: string;
+	postal_code?: string | number;
```
Comment (atanas-dev) [inline client/state/partner-portal/partner/hooks/use-update-company-details.ts]: Let's stick with camelCase when dealing with data inside Calypso so API responses are API* interfaces (e.g. `APIPartner`) which we transform into camelCase interfaces (e.g. Partner) immediately upon receiving them and we flip them to snake_case when we send requests. It's annoying but it also helps to differentiate API data vs internal Calypso data and with being consistent in both environments in terms of naming conventions.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+						onSelect={ ( option: any ) => {
+							setAddressState( option.value );
+						} }
+						disabled={ isSubmitting || ! stateOptions.hasOwnProperty( country ) }
+						isLoading={ Object.keys( stateOptions ).length === 0 }
+					/>
+				) }
+
+				{ ! showCountryFields && <TextPlaceholder /> }
+			</FormFieldset>
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Let's hide this dropdown when it's empty.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+	// The partner has already been fetched through e.g. the requireAccessContext
+	// controller function, so we do not have to do any further checks.
+	const partner = useSelector( getCurrentPartner );
+	const { countryOptions, stateOptions } = useCountriesAndStates();
+	const submitNotificationId = 'company-details-form';
+	const showCountryFields = countryOptions.length > 0;
+
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const [ name, setName ] = useState( castAsString( partner?.name ) );
+	const [ country, setCountry ] = useState( castAsString( partner?.address.country ) );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: The first time I edit this form, if I don't interact with the Country dropdown the state value will remain "" so submitting the form without doing any changes will result in "Invalid parameter(s): country" because it is an empty string.

**Review by atanas-dev — CHANGES_REQUESTED** [review]
The dropdowns can go out of bounds on mobile when closed:
Screen Shot 2022-04-28 at 17 31 26

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts (hunk 2 above)
Comment (kallehauge) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts]: Hey! It's as specific as it gets 😂  Thanks for the tip @atanas-dev 👏

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts (hunk 5 above)
Comment (kallehauge) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts:8]: Yes, just yes 😢

File: client/state/partner-portal/partner/hooks/use-update-company-details.ts (hunk 12 above)
Comment (kallehauge) [reply client/state/partner-portal/partner/hooks/use-update-company-details.ts]: I'd love to get your take on this direction: https://github.com/Automattic/wp-calypso/pull/63100/commits/736ec3ee8bdd7b21b3d16cca8d2361fdf4bb3ab9 (the actual implementation gives TS errors because I'm messing with the interface values. I guess I could work around it by defining the properties as `<unknown>` first 😅 )

When we format the partner, we convert it to `postalCode` so it will be camelCase when we work with it in the code itself, but we keep the payload as `postal_code`.
I kinda like the idea of specifying the exact payload we need to provide + doing it this way we don't have to convert the value in 2 places:
* `mutationUpdateCompanyDetails` as highlighted here.
* When we record the `calypso_partner_portal_update_company_details_submit` track - _the track expect snake and/or kebab case formatting_).

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 14 above)
Comment (kallehauge) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: The ISO 3166 verification in the endpoint didn't allow for empty values. A condition has been added to the diff and should now work as expected 👍

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 10 above)
Comment (kallehauge) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: We now have a `<form>` element and we're disabling the submit button until `showCountryFields` is true.
We're still using controlled components because we're dependent on the value of `country` to show `state`, so we didn't save much. But it's done!

**Comment by kallehauge** [comment]
I have re-ordered some of the fields to better match the i3 "create partner" experience (p1HpG7-fQT-p2)

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 14 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: This is a UI problem, not an API problem. What if the customer is from Afghanistan (the first dropdown option) and they click the submit button without interacting with the dropdown? Their country won't be saved.

Technically we're lying to the user about what the selected value is (state<->UI desync). If we have no selected country (or it doesn't exist in the list of countries) we should set the country value to the first country once we have the list of countries so our state and the UI are always in sync.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 6 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: You might have accidentally missed adjusting this.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 7 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: You might have accidentally missed adjusting this.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+					id: submitNotificationId,
+				} )
+			);
+			setIsSubmitting( false );
+		},
+	} );
+
+	const handleSubmit = useCallback(
+		( e ) => {
+			e.preventDefault();
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx:87]: We should have an early return here to not submit the form if the country list or an update request are still pending:
```js
if ( ! showCountryFields || updateCompanyDetails.isLoading ) {
    return;
}
```

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 8 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: This has been missed.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 9 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: This has been missed.

Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: This has been missed.

File: client/state/partner-portal/partner/hooks/use-update-company-details.ts (hunk 12 above)
Comment (atanas-dev) [reply client/state/partner-portal/partner/hooks/use-update-company-details.ts]: The way I think about this is based on what is internal and what is external.
Anything in Calypso should be in Calypso's format. Anything coming in from the "outside" or going out should be in whatever format the "outside" in that case uses.

As such, JSON coming in from the WPCOM API is in snake_case and when it enters Calypso's domain, it should be converted to an appropriate format for Calypso.
When data needs to go out to the WPCOM API it should be in snake_case as that is what the API uses, even if that means we have to convert it back to snake_case.
The same logic applies to Tracks - it's an "outside" domain and we need to convert it if that is what that domain expects.

For argument's sake, let's say that Stripe used kebab-case and Tracks used MACRO_CASE - which one would we use? Would we allow our Calypso state to be littered with 4 different cases because of where that data originally comes from? Even if that data comes from someplace else the moment it enters Calypso's domain, it becomes Calypso's data and should be treated as such.

Isolating and converting data's structure at the entry and exit points makes the most sense to me, even if that means some extra work on converting the data's structure on the way in and on the way out.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx (hunk 14 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: Ahhh, I just saw there's a "Select country" option so there is an empty option there. My apologies.
In case we want to allow customers to clear their country value we should keep it as-is.
In case we don't want to allow customers to clear their country after they've set it, we should remove the empty option.

The case of a country value being stored on the API side which doesn't exist in the select options still stands though and we should handle it by switching it to the empty option internally (so the internal country state is cleared if the country doesn't exist so submission doesn't keep that invalid country value) if we decide to keep that.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+							} }
+							disabled={ isSubmitting }
+							isLoading={ countryOptions.length === 0 }
+						/>
+					) }
+
+					{ ! showCountryFields && <TextPlaceholder /> }
+				</FormFieldset>
+				<FormFieldset>
+					{ showCountryFields && stateOptions && (
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: There's some extra spacing when the field is not visible because of the fieldset:
Screen Shot 2022-05-19 at 21 03 45

**Comment by atanas-dev** [comment]
Some replies on outdated files are appearing on their own instead of in their threads so I think you'll have to expand the previous set of feedback items in order to see the full context 😩

**Comment by atanas-dev** [comment]
I've taken over the task and have addressed my feedback.

**Review by heydemoura — APPROVED** [review]
LGTM :shipit:

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7331543

Thank you @kallehauge for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7331543).
